contracts/RewardController.sol

```diff
@@ -0,0 +1,252 @@
+// SPDX-License-Identifier: MIT
+// Disclaimer https://github.com/hats-finance/hats-contracts/blob/main/DISCLAIMER.md
+
+pragma solidity 0.8.16;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "./interfaces/IRewardController.sol";
+
+contract RewardController is IRewardController, OwnableUpgradeable {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    // Not enough rewards to transfer to user
+    error NotEnoughRewardsToTransferToUser();
+
+    struct VaultInfo {
+        uint256 rewardPerShare;
+        uint256 lastProcessedTotalAllocPoint;
+        uint256 lastRewardBlock;
+        uint256 allocPoint;
+    }
+
+    struct VaultUpdate {
+        uint256 blockNumber;// update blocknumber
+        uint256 totalAllocPoint; //totalAllocPoint
+    }
+
+    uint256 public constant MULTIPLIERS_LENGTH = 24;
+
+    // Block from which the vaults contract will start rewarding.
+    uint256 public startBlock;
+    uint256 public epochLength;
+    // the ERC20 contract in which rewards are distributed
+    IERC20Upgradeable public rewardToken;
+    // Reward Multipliers
+    uint256[24] public rewardPerEpoch;
+    VaultUpdate[] public globalVaultsUpdates;
+    mapping(address => VaultInfo) public vaultInfo;
+    // vault address => user address => reward debt amount
+    mapping(address => mapping(address => uint256)) public rewardDebt;
+    // vault address => user address => unclaimed reward amount
+    mapping(address => mapping(address => uint256)) public unclaimedReward;
+
+    event SetRewardPerEpoch(uint256[24] _rewardPerEpoch);
+    event SafeTransferReward(
+        address indexed user,
+        address indexed vault,
+        uint256 amount,
+        address rewardToken
+    );
+    event ClaimReward(address indexed _vault, address indexed _user, uint256 _amount);
+
+    function initialize(
+        address _rewardToken,
+        address _hatGovernance,
+        uint256 _startRewardingBlock,
+        uint256 _epochLength,
+        uint256[24] memory _rewardPerEpoch
+    ) external initializer {
+        rewardToken = IERC20Upgradeable(_rewardToken);
+        startBlock = _startRewardingBlock;
+        epochLength = _epochLength;
```

Comment:
lets require that this is > 0, as we divide by this later and it makes me nervous :)